org/mozilla/jss/ssl/javax/JSSEngine.java

```diff
@@ -0,0 +1,334 @@
+    protected SSLVersion max_protocol;
+
+    protected JSSSession session;
+
+    protected boolean is_outbound_closed = false;
+    protected boolean is_inbound_closed = false;
+
+    public JSSEngine() {
+        session = new JSSSession(BUFFER_SIZE);
+    }
+
+    public JSSEngine(String peerHost, int peerPort) {
+        super(peerHost, peerPort);
+
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+    }
+
+    public JSSEngine(String peerHost, int peerPort,
+                     org.mozilla.jss.crypto.X509Certificate localCert,
+                     org.mozilla.jss.crypto.PrivateKey localKey) {
+        super(peerHost, peerPort);
+
+        cert = (PK11Cert) localCert;
+        key = (PK11PrivKey) localKey;
+
+        session = new JSSSession(BUFFER_SIZE);
+        session.setPeerHost(peerHost);
+        session.setPeerPort(peerPort);
+    }
+
+    public void setSSLParameters(SSLParameters params) {
+        JSSParameters parsed;
+
+        // Try to cast the passed parameter into a JSSParameters. This has
+        // additional fields we usually need for construction of a NSS-backed
+        // SSLEngine. Otherwise, parse the values into JSS-specific fields.
+        // If we didn't create the JSSParameters here, both calls
+        // (setEnabledCipherSuites, setEnabledProtocols) would construct their
+        // own JSSParameters.
+        if (params instanceof JSSParameters) {
+            parsed = (JSSParameters) params;
+        } else {
+            parsed = new JSSParameters(params);
+        }
+
+        if (parsed.getSSLCiphers() != null) {
+            setEnabledCipherSuites(parsed.getSSLCiphers());
+        }
+
+        if (parsed.getSSLVersionRange() != null) {
+            setEnabledProtocols(parsed.getSSLVersionRange());
+        }
+
+        setWantClientAuth(parsed.getWantClientAuth());
+        setNeedClientAuth(parsed.getNeedClientAuth());
+
+        // In the event we haven't explicitly set cert and key, try and infer
+        // them from the list of ServerNames specified... We assume that when
+        // the SSLEngine has a certificate already, we want to use them, even
+        // if parsed has a null certificate.
+        if (parsed.getAlias() != null && key_managers != null && key_managers.length > 0 && cert == null && key == null) {
+            setCertFromAlias(parsed.getAlias());
+        }
+
+        if (parsed.getHostname() != null) {
+            setHostname(parsed.getHostname());
+        }
+    }
+
+    public void setHostname(String name) {
+        hostname = name;
+    }
+
+    public void setCertFromAlias(String alias) throws IllegalArgumentException {
+        if (alias == null) {
+            cert = null;
+            key = null;
+            return;
+        }
+
+        if (cert != null || key != null) {
+            throw new IllegalArgumentException("SSLEngine already has certificates; refusing to change them.");
```

Comment:
Hmm.. it looks like you can get around this restriction by calling `setCertFromAlias(null)` first then `setCertFromAlias(newAlias)`.

I think if the cert & key are meant to be immutable it should reject the null alias too. I think it would make sense to throw an exception here so the caller knows that the operation to change the cert/key is rejected.

On the other hand, does it make sense to reuse a JSSEngine instance with a new cert/key? Or is it better to require a new JSSEngine instance?

Reply:
I think a single `SSLEngine` is limited to a single tuple `(server, client)` pair -- you can't necessarily rehandshake with a new client, but you could handshake with the same client again. In general, `SSLEngine`s are single-use, and once that connection terminates, you discard it. SunJSSE doesn't let you resume a closed `SSLEngine`. See: https://github.com/frohoff/jdk8u-dev-jdk/blob/master/src/share/classes/sun/security/ssl/SSLEngineImpl.java#L667-L668